server/util/redact/redact_test.go

```diff
@@ -47,44 +48,92 @@ func TestRedactMetadata_UnstructuredCommandLine_RemovesArgs(t *testing.T) {
 	assert.Equal(t, []string{}, unstructuredCommandLine.Args)
 }
 
-func TestRedactMetadata_StructuredCommandLine_RedactsEnvVars(t *testing.T) {
-	// TODO(bduffany): Migrate env redaction to redactor and enable.
-	t.Skip()
-
+func TestRedactMetadata_StructuredCommandLine_RedactsEnvVarsAndHeadersAndURLSecrets(t *testing.T) {
 	redactor := redact.NewStreamingRedactor(testenv.GetTestEnv(t))
-	shellOption := &command_line.Option{
-		CombinedForm: "--client_env=SHELL=/bin/bash",
-		OptionName:   "client_env",
-		OptionValue:  "SHELL=/bin/bash",
+	// Started event specified which env vars shouldn't be redacted.
+	buildStarted := &bespb.BuildStarted{
+		OptionsDescription: "--build_metadata='ALLOW_ENV=FOO_ALLOWED,BAR_ALLOWED_PATTERN_*'",
 	}
-	secretOption := &command_line.Option{
+	envSecretOption := &clpb.Option{
 		CombinedForm: "--client_env=SECRET=codez",
 		OptionName:   "client_env",
 		OptionValue:  "SECRET=codez",
 	}
-	structuredCommandLine := &command_line.CommandLine{
+	envAllowedByDefaultOption := &clpb.Option{
+		CombinedForm: "--client_env=GITHUB_REPOSITORY=https://username:password@github.com/foo/bar",
+		OptionName:   "client_env",
+		OptionValue:  "GITHUB_REPOSITORY=https://username:password@github.com/foo/bar",
+	}
+	envAllowedByUserOption := &clpb.Option{
+		CombinedForm: "--client_env=FOO_ALLOWED=bar",
+		OptionName:   "client_env",
+		OptionValue:  "FOO_ALLOWED=bar",
+	}
+	envAllowedByUserPatternOption := &clpb.Option{
+		CombinedForm: "--client_env=BAR_ALLOWED_PATTERN_XYZ=qux",
+		OptionName:   "client_env",
+		OptionValue:  "BAR_ALLOWED_PATTERN_XYZ=qux",
+	}
+	remoteHeaderOption := &clpb.Option{
+		CombinedForm: "--remote_header=x-buildbuddy-api-key=bbapikey",
+		OptionName:   "remote_header",
+		OptionValue:  "x-buildbuddy-api-key=bbapikey",
+	}
+	remoteCacheHeaderOption := &clpb.Option{
+		CombinedForm: "--remote_cache_header=x-buildbuddy-api-key=bbapikey",
+		OptionName:   "remote_cache_header",
+		OptionValue:  "x-buildbuddy-api-key=bbapikey",
+	}
+	nonEnvURLOption := &clpb.Option{
+		CombinedForm: "--some_url=https://password@foo.com",
+		OptionName:   "some_url",
+		OptionValue:  "https://password@foo.com",
+	}
+	structuredCommandLine := &clpb.CommandLine{
```

Comment:
Do we get any value out of passing all of the options at once, but making the assertions in a loop? It makes the test harder to read.

Could we do something like the following instead?
option = &clpb.Option{...}
redactor.RedactMetadata(structuredCommandLineEvent(option))
assert.Equal(t, ...)

You could still use a loop to go over the the inputs & expected output combinations if you wanted to.

Reply:
Very much improved -- thanks for the suggestion!